### PR TITLE
arch/sim: Add host timer and improve the oneshot timer logic

### PR DESCRIPTION
**Pull request by sebastianene07**
## Summary of Changes

Add a host timer that generates periodic signals and sends SIGALRM to
the process that runs the NuttX simulation. This logic is integrated as
part of the existing NuttX oneshot timer. The host timer installs an
irq handler which is expected to run every CONFIG_USEC_PER_TICK .

## Impact

This should not impact the current implementation because it adds a new
Kconfig option which is not enabled by default. `CONFIG_SIM_WALLTIME_SIGNAL=n`

## Testing

I verified the implementation by running this config : `sim/ostest`.

Signed-off-by: Sebastian Ene <nuttx@fitbit.com>

### CODE REVIEW

File: arch/sim/Kconfig [hunk 1]
```diff
@@ -509,4 +509,9 @@ config SIM_UART3_NAME
 	default "/dev/ttySIM3"
 	depends on SIM_UART_NUMBER >= 4
 
+config SIM_ONESHOT_HOST_TIMER
+	bool "Use a host timer to deliver oneshot events"
+	default n
+	depends on ONESHOT
```
Comment (xiaoxiang781216) [inline arch/sim/Kconfig:515]: should we change to depend on SIM_WALLTIME? if SIM_WALLTIME equals n, the simulator should run as fast as possible. BTW, is it better to name this option to SIM_WALLTIME_SIGNAL and move after SIM_WALLTIME?

File: arch/sim/src/sim/up_oneshot.c [hunk 2]
```diff
@@ -384,6 +450,19 @@ void up_timer_update(void)
...
+   * timer.
+   */
+
+  if (up_interrupt_context() == false)
+    {
+      return;
+    }
+#endif
+
   clock_timespec_add(&g_current, &tick, &g_current);
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/up_oneshot.c:466]: It's better to add a new static funtion like this:
```
clock_timespec_add(&g_current, &tick, &g_current);
for (entry = sq_peek(&g_oneshot_list); entry; entry = sq_next(entry))
    {
      sim_process_tick(entry);
    }
```
and let's up_timer_update and host_sigalrm_handler call it.

File: arch/sim/src/sim/up_internal.h [hunk 3]
```diff
@@ -232,6 +232,8 @@ void  host_free_shmem(void *mem);
 uint64_t host_gettime(bool rtc);
 void host_sleep(uint64_t nsec);
 void host_sleepuntil(uint64_t nsec);
+int host_setup_timer(void);
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/up_internal.h:235]: How about we let host_setup_timer return(or add a pointer) the irq number?
```
int host_setup_timer(int *irq);
```
so we don't need host_get_alarm_signal.

File: arch/sim/src/sim/up_hosttime.c [hunk 4]
```diff
@@ -89,3 +91,42 @@ void host_sleepuntil(uint64_t nsec)
...
+
+/****************************************************************************
+ * Name: host_setup_timer
+ *
+ * Description:
+ *   Set up a timer to send periodic signals.
+ *
+ ****************************************************************************/
+
+int host_setup_timer(void)
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/up_hosttime.c:103]: Change to host_setimer to align with other function naming?

File: arch/sim/src/sim/up_oneshot.c [hunk 5]
```diff
@@ -299,6 +300,67 @@ static int sim_current(FAR struct oneshot_lowerhalf_s *lower,
...
+ *   events to the OS.
+ *
+ * Input Parameters:
+ *   sig - the signal number
+ *   si  - the signal information
+ *   old_ucontext - the previous context
+ *
+ ****************************************************************************/
+
+static int host_sigalrm_handler(int irq, FAR void *context, FAR void *arg)
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/up_oneshot.c:318]: Rename to sim_alarm_handler?

File: arch/sim/src/sim/up_oneshot.c [hunk 6]
```diff
@@ -299,6 +300,67 @@ static int sim_current(FAR struct oneshot_lowerhalf_s *lower,
...
+ *
+ * Input Parameters:
+ *   nxsched_process_timer - the NuttX scheduler callback
+ *
+ * Returned Value:
+ *   This function returns 0 on success otherwise a negative error code.
+ *
+ ****************************************************************************/
+
+static int host_init_timer(void)
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/up_oneshot.c:338]: Merge the code into up_timer_initialize?

**Review by xiaoxiang781216 — CHANGES_REQUESTED** [review]

File: arch/sim/src/sim/up_oneshot.c (hunk 2 above)
Comment (sebastianene07) [reply arch/sim/src/sim/up_oneshot.c:466]: That's a good idea I will do this

File: arch/sim/src/sim/up_internal.h (hunk 3 above)
Comment (sebastianene07) [reply arch/sim/src/sim/up_internal.h:235]: I will update this, thanks

File: arch/sim/src/sim/up_oneshot.c (hunk 5 above)
Comment (sebastianene07) [reply arch/sim/src/sim/up_oneshot.c:318]: Ok, I will rename it

File: arch/sim/src/sim/up_hosttime.c (hunk 4 above)
Comment (sebastianene07) [reply arch/sim/src/sim/up_hosttime.c:103]: I will update the function name

File: arch/sim/src/sim/up_oneshot.c (hunk 6 above)
Comment (sebastianene07) [reply arch/sim/src/sim/up_oneshot.c:338]: I merged it

File: arch/sim/Kconfig (hunk 1 above)
Comment (xiaoxiang781216) [reply arch/sim/Kconfig:515]: Make sense, but changing to choice may better:
```
choice
	prompt "..."
	default SIM_WALLTIME_xxx

config SIM_WALLTIME

config SIM_WALLTIME_xxx

config SIM_WALLTIME_SIGNAL

endchoice
```

File: arch/sim/src/sim/up_internal.h (hunk 3 above)
Comment (xiaoxiang781216) [reply arch/sim/src/sim/up_internal.h:235]: The sequence isn't important here, it's fine to call host_settimer, irq_attach and irq_enable.

File: arch/sim/src/sim/up_oneshot.c (hunk 2 above)
Comment (xiaoxiang781216) [reply arch/sim/src/sim/up_oneshot.c:466]: Yes, it's wrong to call up_interrupt_context here to check alarm signal because the new signal handler(actually, SIGUSR1 is used for SMP now) may add later.

File: arch/sim/Kconfig (hunk 1 above)
Comment (sebastianene07) [reply arch/sim/Kconfig:515]: Yes I think the choice is more suitable. Should we default it to SIM_WALLTIME_SIGNAL ?

Comment (xiaoxiang781216) [reply arch/sim/Kconfig:515]: No, the default value has to be SIM_WALLTIME_xxx to keep the compatiblity.

File: arch/sim/Kconfig [hunk 7]
```diff
@@ -98,7 +98,16 @@ config SIM_X8664_MICROSOFT
...
-	bool "Execution simulation in near real-time"
+	bool "Execute the simulation in near real-time"
+	default n
+
+if SIM_WALLTIME
+choice
+	prompt "Execute the simulation in near real-time"
+	default SIM_WALLTIME_HOSTSLEEP
+
+config SIM_WALLTIME_HOSTSLEEP
```
Comment (xiaoxiang781216) [inline arch/sim/Kconfig:109]: change to SIM_WALLTIME_SLEEP like SIM_WALLTIME_SIGNAL

File: arch/sim/src/sim/up_oneshot.c [hunk 8]
```diff
@@ -376,25 +437,13 @@ void up_timer_initialize(void)
...
 
   host_sleepuntil(g_current.tv_nsec +
     (uint64_t)g_current.tv_sec * NSEC_PER_SEC);
-#endif
 
-  for (entry = sq_peek(&g_oneshot_list); entry; entry = sq_next(entry))
-    {
-      sim_process_tick(entry);
-    }
+  sim_timer_update();
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/up_oneshot.c:447]: Need call sim_timer_update when CONFIG_SIM_WALLTIME=n
```
#ifdef CONFIG_SIM_WALLTIME_HOSTSLEEP
host_sleepuntil(g_current.tv_nsec +
    (uint64_t)g_current.tv_sec * NSEC_PER_SEC);
#endif
#ifndef CONFIG_SIM_WALLTIME_SIGNAL
sim_timer_update();
#endif
```

File: arch/sim/src/sim/up_simsmp.c [hunk 9]
```diff
@@ -120,7 +120,7 @@ void sched_note_cpu_resume(struct tcb_s *tcb, int cpu);
 static void *sim_idle_trampoline(void *arg)
 {
   struct sim_cpuinfo_s *cpuinfo = (struct sim_cpuinfo_s *)arg;
-#ifdef CONFIG_SIM_WALLTIME
+#ifdef CONFIG_SIM_WALLTIME_HOSTSLEEP
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/up_simsmp.c:123]: don't need change, we need sleep in both case(sleep and signal). Let's revert the change at:
https://github.com/apache/incubator-nuttx/pull/1859/files#diff-d607b8ec5dc28fb1e92e2b4389e2a9a8R106

File: arch/sim/Kconfig [hunk 10]
```diff
@@ -98,7 +98,16 @@ config SIM_X8664_MICROSOFT
 endchoice
 
 config SIM_WALLTIME
-	bool "Execution simulation in near real-time"
+	bool "Execute the simulation in near real-time"
+	default n
+
+if SIM_WALLTIME
+choice
+	prompt "Execute the simulation in near real-time"
```
Comment (xiaoxiang781216) [inline arch/sim/Kconfig:106]: Should we reword the prompt to avoid the dup with line 101?

**Review by xiaoxiang781216 — CHANGES_REQUESTED** [review]

File: arch/sim/Kconfig (hunk 7 above)
Comment (sebastianene07) [reply arch/sim/Kconfig:109]: Ok, I will update it

File: arch/sim/src/sim/up_oneshot.c (hunk 8 above)
Comment (sebastianene07) [reply arch/sim/src/sim/up_oneshot.c:449]: Yes, I saw that (sim_timer_update was not called). Thanks

File: arch/sim/Kconfig (hunk 10 above)
Comment (sebastianene07) [reply arch/sim/Kconfig:106]: Yes, and we should also update the README where SIM_WALLTIME is mentioned.

File: arch/sim/src/sim/up_simsmp.c (hunk 9 above)
Comment (sebastianene07) [reply arch/sim/src/sim/up_simsmp.c:123]: Ok, I will revert this. Out of topic: I also want to integrate the SIGUSR1 to use the NuttX irq handling logic in another patch/PR.

Comment (xiaoxiang781216) [reply arch/sim/src/sim/up_simsmp.c:123]: It's on my TODO list too, it's great that you have time to fix it.

File: arch/sim/src/sim/up_hostirq.c [hunk 11]
```diff
@@ -131,7 +131,7 @@ void up_enable_irq(int irq)
   memset(&act, 0, sizeof(act));
   act.sa_sigaction = up_handle_irq;
   act.sa_flags     = SA_SIGINFO;
-  sigfillset(&act.sa_mask);
+  sigemptyset(&act.sa_mask);
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/up_hostirq.c:134]: Why need this change?

Comment (sebastianene07) [reply arch/sim/src/sim/up_hostirq.c]: We don't need it. I will revert it.

**Review by xiaoxiang781216 — COMMENTED** [review]
@sebastianene07 there are still several comments not addressed.

**Comment by sebastianene07** [comment]
> @sebastianene07 there are still several comments not addressed.

Correct, sorry. I forgot to push

**Comment by xiaoxiang781216** [comment]
@sebastianene07 need move code in arch/sim/src/Makefile:
```
  HOSTCFLAGS += -DCONFIG_USEC_PER_TICK=$(CONFIG_USEC_PER_TICK)
```
from line 104 to line 85.

**Comment by sebastianene07** [comment]
> @sebastianene07 need move code in arch/sim/src/Makefile:
> 
> ```
>   HOSTCFLAGS += -DCONFIG_USEC_PER_TICK=$(CONFIG_USEC_PER_TICK)
> ```
> 
> from line 104 to line 85.

Done, thanks

**Review by xiaoxiang781216 — APPROVED** [review]
LGTM, does anyone have more comment? If not, I will merge the change in tomorrow.
